MARP-149 Standardization & professionalization of Github documentation

nqhoan-axonivy opened the pull request:
Introduce new script for migrating the CODE_OF_CONDUCT.md and SECURITY.md to all repo

nqhoan-axonivy commented:
Hi @weissreto could you help review and give your feedback for this PR.
Thank you so much for your help!

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/GitHubFileDetector.java:
```diff
@@ -0,0 +1,110 @@
+import org.kohsuke.github.GHContent;
+import org.kohsuke.github.GHRepository;
+
+public abstract class GitHubFileDetector {
```
In this repo the standard indent is two spaces. Please replace all /t to two spaces "  ".

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/GitHubMissingCodeOfConductDetector.java:
```diff
@@ -0,0 +1,34 @@
+
+public class GitHubMissingCodeOfConductDetector extends GitHubFileDetector {
+
+	public static void main(String[] args) throws IOException {
```
In which build is this class called?
You should call it in a build like here:
https://github.com/axonivy/github-repo-manager/tree/master/build/public-repo-license

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/GitHubMissingCodeOfConductDetector.java:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	@Override
+	protected String getRequiredFilePath() {
```
I would like to see only one main class and one build for all three files LICENCE, SECURITY, CODE_OF_CONDUCT.
Instead of using inheritance for each file, I would prefer that you provide a list of constant objects to the main class. In the constant object you can provide the meta information for each file. You could use a record like
public record(String requiredFilePath, String pullRequestTitle, String branchName, String commitMessage) {}

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/GitHubFileDetector.java:
```diff
@@ -0,0 +1,110 @@
+package com.axonivy.github;
```
Since all your new classes and GitHubMissingLicenseDetector belong together, I would put them into a new package com.axonivy.github.file

weissreto reviewed (changes requested):
Thanks for the PR. 
Please change \t to spaces.
Only one main class GitHubMissingFiles that is also called from the build. 
Rename the build to public-repo-missing-files
Instead of inheritance use a list of constants. A constant for each file.

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/GitHubFileDetector.java:
```diff
@@ -0,0 +1,110 @@
+		}
+	}
+
+	protected void printMessage(String pattern, Object... arguments) {
```
Please print information to System.out and errors to System.err. See GitHubMissingLicenseDetector

weissreto reviewed (changes requested):
Use System.out for information messages.

nqhoan-axonivy commented:
Thank you so much, @weissreto!
I handled your feedback and clean up something.
But I'm not sure I write on "remove class License detector" and create a list of constant.
Does my change is correct?

My Test result:
![image](https://github.com/axonivy/github-repo-manager/assets/83745591/252e554b-6d87-4d99-a8d7-5c0e6a81a607)

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/file/GitHubMissingFiles.java:
```diff
@@ -0,0 +1,35 @@
+      "Add Apache 2.0 license");
+  private static final FileMeta SECURITY = new FileMeta("SECURITY.md", "Add security file", "Add_Security",
+      "Add Security.md file to repo");
+  private static final FileMeta CODE_OF_CONDUCT = new FileMeta("CODE_OF_CONDUCT.md", "Add code of conduct file",
```
create a constant for List.of(LICENSE, SECURITY, CODE_OF_CONDUCT); 
the iterate over the items in the list in the main method.

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/file/GitHubMissingFilesDetector.java:
```diff
@@ -0,0 +1,118 @@
+    }
+  }
+
+  private boolean hasSimilarFilePaths(GHContent existingFile) {
```
This method does not make sense anymore. It was here because getLicense() returned licenses in LICENSE and COPYING file. Since you now work with getFileContent() this check no longer makes sense.

Instead, we may want to check if the content of the file is the same as the standard one?

weissreto reviewed (changes requested):
Getting better. Some more comments below.

nqhoan-axonivy commented:
Hi @weissreto I updated the code according to your suggestion, and changed the method hasSimilarFilepath to compare the content of files, I tested and found that it worked for me when I have the same filename but different content.
And, I also introduced a new class - GitHubFiles, I would like to move the described required files to a separate class, then in the future, if we want to add more files, we can extend this class and not make the main class hard to read - if this idea is not make sense, just feedback, I can revert it quickly. Thanks!

weissreto commented on github-repo-manager/src/main/java/com/axonivy/github/file/GitHubMissingFiles.java:
```diff
@@ -0,0 +1,27 @@
+    for (var fileMeta : REQUIRED_FILES) {
+      var githubMissingFiles = new GitHubMissingFilesDetector(fileMeta, WORKING_ORGANIZATIONS);
+      var returnedStatus = githubMissingFiles.checkMissingFile();
+      missingStatus = returnedStatus != 0 ? returnedStatus : 0;
```
If returnedStatus is not 0 in the first item and 0 in the second the result is 0 which is wrong. Code should be:
missingStatus = returnedStatus != 0? returnedStatus : missingStatus;

nqhoan-axonivy replied:
ah ah sorry Reto, my bad.
I updated it, I have the same idea but I described it wrong.

weissreto reviewed (changes requested):
One small comment that you need to fix before I merge

weissreto commented:
The following is the output of the build. Can you have a look at it?
[build.txt](https://github.com/axonivy/github-repo-manager/files/15075070/build.txt)

nqhoan-axonivy commented:
> The following is the output of the build. Can you have a look at it? [build.txt](https://github.com/axonivy/github-repo-manager/files/15075070/build.txt)

Thank @weissreto!
I checked the log and 99% is as I expected,
Just one concern on the last statement:
ERROR: script returned exit code 255
[GitHub Checks] GitHub check (name: Jenkins, status: completed) has been published.
GitHub has been notified of this commit’s build result
Finished: FAILURE
=> So it means the java code gets issue or am I missing something in maven command line?

weissreto commented:
Hi @nqhoan-axonivy 

Seems you have not adjusted the Code of Conduct? It contains mail addresses and links to a strange web site. I think this should be adjusted for Axon Ivy.

ivy-rew commented:
Why did we introduce this code-of-conduct to all repositories of axonivy?
IMHO it doesn't make us look professional, exactly the opposite if we don't talk about it and just add some generic COC. 
I'd strongly vote to remove it, and talk about its content before just polluting each and every repo with it.
